File: 19. Remove Nth Node From End of List.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        first, second = dummy, dummy

        for i in range(n + 1):
            first = first.next

        while first:
            first = first.next
            second = second.next

        second.next = second.next.next if second.next else second.next

        return dummy.next

File: cpp/linked-list/19. Remove Nth Node From End of List.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {   
        ListNode dummy = ListNode(0,head), *prev = &dummy, *curr = &dummy; 

        for (int i = 0; i <= n; ++i) {
            curr = curr->next;
        }
        while(curr) {
            prev = prev->next;
            curr = curr->next;
        }
        prev->next = prev->next->next; 
        return dummy.next; 
    }
};